typing/ctype.ml

```diff
@@ -4884,25 +4884,63 @@ let same_constr env t1 t2 =
 let () =
   Env.same_constr := same_constr
 
-let is_immediate = function
-  | Type_immediacy.Unknown -> false
-  | Type_immediacy.Always -> true
-  | Type_immediacy.Always_on_64bits ->
-      (* In bytecode, we don't know at compile time whether we are
-         targeting 32 or 64 bits. *)
-      !Clflags.native_code && Sys.word_size = 64
-
-let immediacy env typ =
-   match (repr typ).desc with
-  | Tconstr(p, _args, _abbrev) ->
-    begin try
-      let type_decl = Env.find_type p env in
-      type_decl.type_immediate
-    with Not_found -> Type_immediacy.Unknown
-    (* This can happen due to e.g. missing -I options,
-       causing some .cmi files to be unavailable.
-       Maybe we should emit a warning. *)
+(* We use expand_head_opt version of expand_head to get access
+   to the manifest type of private abbreviations. *)
+let rec get_unboxed_type_representation env ty fuel =
```

Comment:
Why did you move this (not-so-nice) code from its own module to `Ctype`? I suppose there are dependency reasons, but it's not so clear what they are. I would have a preference for having the logic (for the unboxed stuff, and for the immediacy stuff) in their own modules instead of a file of 5K lines.

Reply:
Yes, circular dependency reasons. I'll try to remember what they were.

Reply:
I think that it might be worth merging this functionality into `try_expand_once_opt`. The original aim of that function was to "expand" a type to get another type with the same representation. The intention is to use this for representation optimisations -- the "opt" in the name stands for optimisation -- like specializing primitives. That use case would be better served by a function that went through unboxed types too.

You'll need to be a little careful making this change as IIRC there a couple of uses of `type_expand_once_opt` that really only want to expand through `private` expansions, they would need to use a function with the same behaviour as the current `type_expand_once_opt` called e.g. `type_expand_once_private`.